File: wave2_network/packages/shared-py/blindference_utils/ipfs.py
from __future__ import annotations

import os

import requests


def _api_key() -> str:
    api_key = os.getenv("LIGHTHOUSE_API_KEY")
    if not api_key:
        raise RuntimeError("LIGHTHOUSE_API_KEY is not set")
    return api_key


def upload_to_ipfs(data: bytes) -> str:
    response = requests.post(
        "https://node.lighthouse.storage/api/v0/add",
        files={"file": ("blob", data)},
        headers={"Authorization": f"Bearer {_api_key()}"},
        timeout=60,
    )
    response.raise_for_status()

    payload = response.json()
    cid = payload.get("Hash")
    if not cid:
        raise RuntimeError(f"IPFS upload failed: {payload}")

    return str(cid)


def download_from_ipfs(cid: str) -> bytes:
    response = requests.get(
        f"https://gateway.lighthouse.storage/ipfs/{cid}",
        timeout=60,
    )
    response.raise_for_status()
    return response.content

File: wave2_network/packages/shared/src/ipfs.ts
import lighthouse from '@lighthouse-web3/sdk'

function getApiKey(): string {
  const apiKey = process.env.LIGHTHOUSE_API_KEY
  if (!apiKey) {
    throw new Error('LIGHTHOUSE_API_KEY is not set')
  }
  return apiKey
}

export async function uploadToIPFS(data: Buffer): Promise<string> {
  const response = await lighthouse.uploadBuffer(data, getApiKey())

  if (!response.data?.Hash) {
    throw new Error(`IPFS upload failed: ${JSON.stringify(response)}`)
  }

  return response.data.Hash
}

export async function downloadFromIPFS(cid: string): Promise<Buffer> {
  const url = `https://gateway.lighthouse.storage/ipfs/${cid}`
  const response = await fetch(url)

  if (!response.ok) {
    throw new Error(`IPFS download failed: ${response.status} ${response.statusText}`)
  }

  const arrayBuffer = await response.arrayBuffer()
  return Buffer.from(arrayBuffer)
}
